### PR TITLE
client-cert: certificate chain in the same order as it appears in TLS

### PR DESCRIPTION
**Pull request by bc-pi**
attempt @martinthomson's "simple alternative" in #1927 to say the chain is in the same order as it appears in TLS rather than copying the language from TLS

### CODE REVIEW

File: draft-ietf-httpbis-client-cert-field.md [hunk 1]
```diff
@@ -183,11 +183,13 @@ multiple times in a request.
...
 In the context of a TLS terminating reverse proxy deployment, the proxy
-MAY make the certificate chain used for validation of the end-entity certificate
+MAY make the certificate chain for validation of the end-entity certificate
 available to the backend application with the Client-Cert-Chain HTTP header
-field. This field contains certificates used by the proxy to validate the
-certificate used by the client in the TLS handshake. These certificates might or
-might not have been provided by the client during the TLS handshake.
+field. This field contains certificates for validation of the end-entity
+certificate provided by the client in the TLS handshake. These certificates might or
+might not have been provided by the client during the TLS handshake, however,
```
Comment (martinthomson) [inline draft-ietf-httpbis-client-cert-field.md:190]: Can you explain your reasoning behind this "might or might not" clause?

Comment (bc-pi) [reply draft-ietf-httpbis-client-cert-field.md:190]: That text was from 0a7210f7d644dea399f0fc4d1b5119acb436d1db when @MikeBishop added the Client-Cert-Chain header and I was trying not to step on his work too much. And also to allow for the case that the proxy assembles the chain and conveys it.

Comment (martinthomson) [reply draft-ietf-httpbis-client-cert-field.md:190]: This seems problematic as it invites a query about why that might (or might not) be the case.

I would prefer to say that the proxy can add certificates to this field.  It provides the certificates in the order in which they appear in the TLS handshake, unless it has cause to provide other information.

Comment (bc-pi) [reply draft-ietf-httpbis-client-cert-field.md:190]: What about even less invitation of questions by making that last sentence this? 

"The order of the certificates is the same as the ordering in TLS (such as described in {{Section 4.4.2 of RFC8446}})."

Comment (martinthomson) [reply draft-ietf-httpbis-client-cert-field.md:190]: That would work for me, but I got the impression that you are aware of cases where that is not the case.  I don't think that there is any harm in saying "..., unless the server and proxy agree otherwise." or similar.

Comment (MikeBishop) [reply draft-ietf-httpbis-client-cert-field.md:190]: Honestly don't remember what my logic was in writing that; it appears to have been from my first pass at this text and not touched in any of the feedback.

I *think* that was based on TLS 1.3 saying that "...a certificate that specifies a trust anchor MAY be omitted from the chain, provided that supported peers are known to possess any omitted certificates."  If the proxy possesses certificates the origin might not, the client might omit them from the handshake, but the proxy could append them for the origin's information.

I don't think there's harm in allowing for private agreement here, but I don't feel strongly either way.

Comment (bc-pi) [reply draft-ietf-httpbis-client-cert-field.md:190]: As I remember things, there was some disagreement when the initial text for `Cient-Cert-Chain` was added/proposed about whether the proxy would construct and validate the chain and convey the constructed chain to the origin or whether the proxy would pass through what the client sent as-is and defer validation and construction to the origin. Some of this discussion is captured in PR #1874. AFAIK most of the time the chain would look the same. But I've tried to adjust the text such that either would be okay. The sentence above about the order being the same was intended to mean only that ordering is the same but not that they are the same.

Comment (b---c) [reply draft-ietf-httpbis-client-cert-field.md]: [5fb237a](https://github.com/httpwg/http-extensions/pull/2258/commits/5fb237a7dbb030ab323710a1d254d442cd41f35d) is another pass at this that removes the "might or might not" and tries to articulate ordering.

File: draft-ietf-httpbis-client-cert-field.md [hunk 2]
```diff
@@ -183,15 +183,13 @@ multiple times in a request.
...
 available to the backend application with the Client-Cert-Chain HTTP header
-field. This field contains certificates used by the proxy to validate the
-certificate used by the client in the TLS handshake. These certificates might or
-might not have been provided by the client during the TLS handshake.
+field.
 
 Client-Cert-Chain is a List Structured Header {{!RFC8941}}.  Each item in the
 list MUST be a Byte Sequence ({{Section 3.3.5 of RFC8941}}) encoded as described
-in {{encoding}}.
+in {{encoding}}. The order is the same as the ordering in TLS (such as described in {{Section 4.4.2 of RFC8446}})
```
Comment (martinthomson) [inline draft-ietf-httpbis-client-cert-field.md:192]:
```suggestion
in {{encoding}}. The order is the same as the ordering in TLS (such as described in {{Section 4.4.2 of RFC8446}}).
```

File: draft-ietf-httpbis-client-cert-field.md [hunk 3]
```diff
@@ -142,7 +142,7 @@ the reverse proxy from the reverse proxy to the origin server.
 
 Client-Cert-Chain:
 : Conveys the certificate chain used for validation of the end-entity
-certificate used by the client in the TLS handshake from the reverse proxy to the origin server.
+certificate provided by the client in the TLS handshake with the reverse proxy from the proxy to the origin server.
```
Comment (martinthomson) [inline draft-ietf-httpbis-client-cert-field.md:145]: This sentence has gotten a little hard to follow.

There are two things here that might need to be separated to make this more readable:

1. This is the chain (less the EE cert) that the client presents in TLS to the TTRP.
2. The TTRP sends this information to the origin server.

File: draft-ietf-httpbis-client-cert-field.md [hunk 4]
```diff
@@ -506,12 +496,13 @@ latter and ensure wide applicability by not trying to cherry-pick particular
...
-The handshake and validation of the client certificate (chain) of the
-mutually-authenticated TLS connection is performed by the TTRP.  With the
+The validation of the client certificate and chain of the mutually-authenticated
+TLS connection is typically performed by the TTRP during the handshake.  With the
 responsibility of certificate validation falling on the TTRP, the
 end-entity certificate is oftentimes sufficient for the needs of the origin server.
 The separate `Client-Cert-Chain` field can convey the certificate chain for
-deployments that require such information.
+deployments that require such information in order for the origin server to augment
+the chain validation or assume it entirely.
```
Comment (martinthomson) [inline draft-ietf-httpbis-client-cert-field.md:505]:
```suggestion
origin server deployments that require this additional information.
```

Maybe simpler is better.

**Review by martinthomson — COMMENTED** [review]
This works well enough on the critical points.  Some points about readability though.

File: draft-ietf-httpbis-client-cert-field.md (hunk 4 above)
Comment (b---c) [reply draft-ietf-httpbis-client-cert-field.md:505]: Maybe so.

File: draft-ietf-httpbis-client-cert-field.md (hunk 3 above)
Comment (b---c) [reply draft-ietf-httpbis-client-cert-field.md:145]: I know. I know. Saying it is "a little hard to follow" is being too kind. It's pretty terrible. And I seem to make it worse each time I touch it. I'll take another pass over it though with that separation in mind. Maybe leaning more on the text above to make the second point and leaving it out of the Client-Cert/Client-Cert-Chain items.
